### ISSUE TITLE
Sik-Ho Tang | Review: MobileNetV1 -- Depthwise Separable Convolution (Light Weight Model).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#114]
Sik-Ho Tang. [Review: MobileNetV1 — Depthwise Separable Convolution (Light Weight Model)](https://towardsdatascience.com/review-mobilenetv1-depthwise-separable-convolution-light-weight-model-a382df364b69).

**Comment by NorbertZheng** [comment]
## Overview
In this story, MobileNetV1 from Google is reviewed. Depthwise Separable Convolution is used to reduce the model size and complexity. It is particularly useful for mobile and embedded vision applications.
 - Smaller model size: Fewer number of parameters.
 - Smaller complexity: Fewer Multiplications and Additions (Multi-Adds).

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/45550351-b324-4dcc-a01d-c8bcdffcfbf3)
When MobileNets Applied to Real Life.

Two parameters are introduced so that MobileNet can be tuned easily: Width Multiplier $\alpha$ and Resolution Multiplier $\rho$. And this is a paper published in arXiv 2017 [1] with more than 600 citations when I was writing this paper.

The above object detection example is the MobileNet which is actually amazing because it can achieve around 25 fps with such large amount of objects detected at the same time.

**Comment by NorbertZheng** [comment]
## Depthwise Separable Convolution
Depthwise separable convolution is a depthwise convolution followed by a pointwise convolution as follows:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/4a9b0075-a8ad-4fae-aa75-98a9eb77fb44)
 - Depthwise convolution is the channel-wise $D_{K}\times D_{K}$ spatial convolution. Suppose in the figure above, we have 5 channels, then we will have 5 $D_{K}\times D_{K}$ spatial convolution.
 - Pointwise convolution actually is the $1\times 1$ convolution to change the dimension.

With above operation, the operation cost is:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/641ea058-68e4-4708-8873-d0612382e14f)
Depthwise Separable Convolution Cost: Depthwise Convolution Cost (Left), Pointwise Convolution Cost (Right).

where $M$: Number of input channels, $N$: Number of output channels, $D_{K}$: Kernel size, and $D_{F}$: Feature map size.

For standard convolution, it is:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/a132bb6e-c6bd-4453-bc4a-08e75b8cbd52)
Standard Convolution Cost.

Thus, the computation reduction is:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/5f80300b-2240-4cc9-8401-54ad90dc985e)
Depthwise Separable Convolution Cost / Standard Convolution Cost.

When $D_{K}\times D_{K}$ is $3\times 3$, 8 to 9 times less computation can be achieved, but with only small reduction in accuracy.

**Comment by NorbertZheng** [comment]
## Whole Network Architecture
Below is the MobileNet Architecture:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/9c596d29-3e28-4bff-be40-58748f66f312)
Whole Network Architecture for MobileNet.

It is noted that Batch Normalization (BN) and ReLU are applied after each convolution:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d9cd0c7c-b740-4e47-beac-74ef318f7622)
Standard Convolution (Left), Depthwise separable convolution (Right) With BN and ReLU.

If Standard Convolution vs Depthwise Separable Convolution for ImageNet dataset:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c92f66ad-6d48-4554-849a-614682317c90)
Standard Convolution vs Depthwise Separable Convolution (ImageNet dataset).

MobileNet only got 1% loss in accuracy, but the Mult-Adds and parameters are reduced tremendously.

**Comment by NorbertZheng** [comment]
## Width Multiplier $\alpha$ for Thinner Models
Width Multiplier $\alpha$ is introduced to control the number of channels or channel depth, which makes $M$ become $\alpha M$. And the depthwise separable convolution cost become:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/653fe024-7f2a-4f14-aded-804f0d45dbc0)
Depthwise Separable Convolution Cost with Width Multiplier $\alpha$.

where $\alpha$ is between 0 to 1, with typical settings of 1, 0.75, 0.5 and 0.25. When $\alpha=1$, it is the baseline MobileNet. And the computational cost and the number of parameters can be reduced quadratically by roughly $\alpha^{2}$.
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c5799cf3-58ec-4ba3-93dc-195ac802db81)
Different Values of Width Multiplier $\alpha$.

Accuracy drops off smoothly from $\alpha=1$ to $0.5$ until $\alpha=0.25$ which is too small.

**Comment by NorbertZheng** [comment]
## Resolution Multiplier $\rho$ for Reduced Representation
Resolution Multiplier $\rho$ is introduced to control the input image resolution of the network, with Resolution Multiplier $\rho$, the cost become:

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/a25cbf64-2fcd-4c3b-822c-00191dc503ef)
Depthwise Separable Convolution Cost with Both Width Multiplier $\alpha$ and Resolution Multiplier $\rho$.

where $\rho$ is between $0$ to $1$. And the input resolution is 224, 192, 160, and 128. When $\rho=1$, it is the baseline MobileNet.
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/fab4197f-ce2e-47d2-906d-ccdb56ea68e1)

Different Values of Resolution Multiplier $\rho$.

Accuracy drops off smoothly across resolution from 224 to 128.

**Comment by NorbertZheng** [comment]
## Comparison With State-of-the-art Approaches
When 1.0 MobileNet-224 is used, it outperforms GoogLeNet (Winner of ILSVRC 2014) and VGGNet (1st Runner Up of ILSVRC 2014) while the multi-adds and parameters are much fewer:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/42ff65a5-2b21-46ea-87b1-b1af5b21861a)
ImageNet Dataset.

When smaller network, 0.50 MobileNet-160, is used, it outperforms Squeezenet and AlexNet (Winner of ILSVRC 2012) while the multi-adds and parameters are much fewer:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/90190fb1-65b2-47d0-93bd-b6d013e3c6f5)
ImageNet Dataset.

It is also competitive with Inception-v3 (1st Runner Up in ILSVRC 2015) while the multi-adds and parameters are much fewer:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/18dc6ecb-dbae-43b6-ae2b-2e67615904fc)
Stanford Dogs Dataset.

**Comment by NorbertZheng** [comment]
Many other datasets are also tried to prove the effectiveness of MobileNet:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/2271205c-762a-4046-a63e-078ec7218fb5)
GPS Localization Via Photos.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d32316db-4a85-469d-b1f9-ebb8c57a96cc)
Face Attribute Classification.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/6f552072-0e31-4e88-8bf1-3854be5c1d27)
Microsoft COCO Object Detection Dataset.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/649b2e15-9a58-4bff-9ed0-92bd8e429760)
MobileNet SSD.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/57f4a8f7-44a7-4e42-9707-2e6dfb2bdc61)
Face Recognition.

To conclude, similar performance with state-of-the-art approaches but with much smaller network is achieved using MobileNet, favored by Depthwise Separable Convolution.

Indeed, there are still many applications I haven’t mentioned above, like GPS Localization Via Photos, Face Attribute Classification and Face Recognition. Hope I can cover them in the coming future. :)

**Comment by NorbertZheng** [comment]
## References
 - [2017 arXiv] [MobileNetV1] [MobileNets: Efficient Convolutional Neural Networks for Mobile Vision Applications](https://arxiv.org/abs/1704.04861).
